drv/lpc55-swd/src/main.rs

```diff
@@ -484,16 +488,50 @@ impl idl::InOrderSpCtrlImpl for ServerImpl {
             Err(_) => Err(SpCtrlError::Fault.into()),
         }
     }
+
+    fn enable_sp_slot_watchdog(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+        time_ms: u32,
+    ) -> Result<(), RequestError<SpCtrlError>> {
+        ringbuf_entry!(Trace::EnabledWatchdog);
+        if !self.init {
+            return Err(SpCtrlError::NeedInit.into());
+        }
+        // This function is idempotent(ish), so we don't care if the timer was
+        // already running; set the new deadline based on current time.
+        let deadline = sys_get_timer().now + time_ms as u64;
+        sys_set_timer(Some(deadline), notifications::TIMER_MASK);
+        Ok(())
+    }
+
+    fn disable_sp_slot_watchdog(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+    ) -> Result<(), RequestError<core::convert::Infallible>> {
+        ringbuf_entry!(Trace::DisabledWatchdog);
+        sys_set_timer(None, notifications::TIMER_MASK);
+        Ok(())
+    }
 }
 
 impl NotificationHandler for ServerImpl {
     fn current_notification_mask(&self) -> u32 {
-        // We don't use notifications, don't listen for any.
-        0
+        notifications::TIMER_MASK
     }
 
     fn handle_notification(&mut self, _bits: u32) {
-        unreachable!()
+        ringbuf_entry!(Trace::WatchdogFired);
+
+        // Disable the watchdog timer
+        sys_set_timer(None, notifications::TIMER_MASK);
+
+        // Attempt to do the swap
+        let r = self.swap_sp_slot();
+        ringbuf_entry!(Trace::WatchdogSwap(r));
+
+        // Force reinitialization
+        self.init = false;
     }
```

Comment:
I don't think we need to be concerned about introducing asynchronicity here based on how hubris has been designed but I'm still a little wary here. Have you tested with how this interacts with the `sp_measure` task? That task is mostly for reference as full measurement is still WIP but I think we at least need to consider that case.

Reply:
It looks like `sp_measure` sends RPC messages to `drv-lpc55-swd`, so I think we're fine from a concurrency perspective: while `swd` is in this function, those notifications will just be pending (and we'll get to them when we return to the main loop).

However, this question made me think about handling of the reset pin: it's (as of this PR) owned by `drv-lpc55-swd`, configured as an output, and used to reset the SP.  At some point, `sp_measure` may want to own it instead, configure it as an input, and use it to detect SP resets.  Should `swd` be resetting the SP by poking `AIRCR` instead?

(this was the case before 042d8292c5151bff8b5171e10abe46ef8efcd6e4; the downside is that poking `AIRCR` is fallible)

If we keep it as an output, it should be open-drain instead of push-pull, to avoid injecting current when the SP resets itself; fixed in 794ad434f628454c5122c963b70705926cc46bd1